src/commands/deployCommands/clean.js

```diff
@@ -0,0 +1,79 @@
+const apiMiddleware = require('../deployUtils/apiMiddleware');
+
+const stopOldVersions = (serving, api, versions) =>
+	api
+		.allocations()
+		.then(allocs => {
+			const runningVersions = versions.filter(
+				v => v.servingStatus === 'SERVING'
+			);
+			const servingIds = Object.keys(allocs);
+			console.log('Currently serving traffic:', servingIds.join(', '));
+			console.log(
+				'Currently available to serve:',
+				runningVersions.map(({ id }) => id).join(', ')
+			);
+			const oldestAllocIndex = runningVersions.reduce(
+				(oldestI, v, i) => (servingIds.includes(v.id) ? i : oldestI),
+				-1
+			);
+
+			const stopIndex = Math.max(oldestAllocIndex + serving, 0);
+			const versionsToStop = runningVersions.slice(stopIndex);
+			console.log(
+				'Stopping version(s):',
+				versionsToStop.map(({ id }) => id).join(', ')
+			);
+			// return the versions that should be stopped
+			return versionsToStop;
```

Comment:
SO to describe what I think is going on here. :) It's going through the list of `versions` that are "SERVING" and finding the oldest ones, but making sure that the min number of running are still up. But, shouldn't we also be checking that those versions have `0` traffic? Maybe that is in here, and i'm missing that part.

Reply:
I'll add more line-by-line comments because there are a few steps to follow:

1. `oldestAllocIndex` is the index of the 'oldest' version that has allocated production traffic - we can be confident that _no versions with a higher array index_ are serving traffic. We then stop everything after `oldestAllocIndex + minimum serving` - it's very possible there are some versions that are younger/newer than `oldestAllocIndex`, but they won't be counted when determining where to begin stopping. Does that make sense?

If the 'serving array' looks like this:

```
v9 (newest) - v8 - v6 - v5 - v4 - v3 - v2 -v1 (oldest)
```

and `v6` is serving traffic and the 'minimum # serving' is 2, then this routine will stop `v4`, `v3`, `v2`, `v1`.

Reply:
This is essentially a `stopSafe` function, similar to `deleteSafe`, but we have to add some extra logic in so that we don't stop anything _newer_ than the _oldest_ version serving prod traffic

Reply:
Excellent! Yeah added comments are beneficial as it's a lot of logic to mentally process. :)